File: python/cryptology/classical/transposition/simple/scytale.py
"""
Scytale (Spartan Stick) Cipher implementation.

The Scytale cipher is one of the oldest known transposition ciphers,
used by the ancient Spartans around 400 BC. It involves wrapping a
leather strip around a cylindrical rod, writing the message along
the length of the rod, and then unwrapping it to reveal the cipher.

Modern implementation uses a matrix approach where we:
1. Write text in rows based on key (wrapping around rod)
2. Read it column by column (unwrapping from rod)

Args:
    text: Plaintext or ciphertext to encrypt/decrypt
    key: The diameter of the rod (number of columns)
    encrypt: True for encryption, False for decryption

The key determines how many columns to use, which corresponds
to the diameter of the cylindrical rod used in ancient times.
"""


def encrypt(plaintext: str, key: int) -> str:
    """
    Encrypt plaintext using Scytale cipher.
    
    Args:
        plaintext: Text to encrypt (case-insensitive, removes non-alphabetic)
        key: Number of columns (diameter of the rod)
    
    Returns:
        Encrypted ciphertext
    
    Example:
        >>> encrypt("HELLO WORLD", 3)
        'HLWL OEOLRDL'
        >>> encrypt("CRYPTOGRAPHY", 4)
        'CYY ARG THPROG P'
    """
    if key <= 0:
        raise ValueError("Key must be positive")
    
    # Clean and normalize text to lowercase
    text = ''.join([c.lower() for c in plaintext if c.isalpha()])
    
    if not text:
        return ""
    
    # Calculate number of rows
    rows = (len(text) + key - 1) // key  # Ceiling division
    
    # Create matrix: write text row by row, filling with spaces if needed
    matrix = []
    text_index = 0
    
    for row in range(rows):
        row_text = []
        for col in range(key):
            if text_index < len(text):
                row_text.append(text[text_index])
                text_index += 1
            else:
                row_text.append(' ')  # Padding with space
        matrix.append(row_text)
    
    # Read column by column (unwrapping from rod)
    result = []
    for col in range(key):
        for row in range(rows):
            result.append(matrix[row][col])
    
    return ''.join(result).strip()


def decrypt(ciphertext: str, key: int) -> str:
    """
    Decrypt ciphertext using Scytale cipher.
    
    Args:
        ciphertext: Encrypted text
        key: Number of columns (diameter of the rod)
    
    Returns:
        Decrypted plaintext
    
    Example:
        >>> decrypt("HLWL OEOLRDL", 3)
        'HELLOWORLD'
    """
    if key <= 0:
        raise ValueError("Key must be positive")
    
    # Clean the ciphertext to lowercase
    text = ciphertext.lower()
    
    if not text:
        return ""
    
    # Reverse the process: read column by column, write row by row
    rows = len(text) // key
    if len(text) % key != 0:
        rows += 1
    
    # Create matrix: write text column by column
    matrix = [[' ' for _ in range(key)] for _ in range(rows)]
    text_index = 0
    
    for col in range(key):
        for row in range(rows):
            if text_index < len(text):
                matrix[row][col] = text[text_index]
                text_index += 1
    
    # Read row by row
    result = []
    for row in range(rows):
        for col in range(key):
            result.append(matrix[row][col])
    
    return ''.join(result).strip()


def get_key_range(text_length: int) -> tuple:
    """
    Get the valid range of keys for a given text length.
    
    Args:
        text_length: Length of the text
    
    Returns:
        Tuple of (min_key, max_key)
    """
    if text_length <= 0:
        return (0, 0)
    return (1, text_length)


if __name__ == "__main__":
    # Test the implementation
    print("=" * 60)
    print("SCYTALE CIPHER - Spartan Stick Cipher")
    print("=" * 60)
    
    plaintext = "HELLO WORLD"
    key = 3
    
    print(f"\nPlaintext:  {plaintext}")
    print(f"Key:        {key}")
    print(f"Plaintext (cleaned): {''.join([c.lower() for c in plaintext if c.isalpha()])}")
    
    encrypted = encrypt(plaintext, key)
    decrypted = decrypt(encrypted, key)
    
    print(f"\nEncrypted:  {encrypted}")
    print(f"Decrypted:  {decrypted}")
    print(f"Success:    {decrypted == ''.join([c.lower() for c in plaintext if c.isalpha()])}")
    
    # Test with different key
    print("\n" + "=" * 60)
    plaintext2 = "CRYPTOGRAPHY"
    key2 = 4
    
    print(f"\nPlaintext:  {plaintext2}")
    print(f"Key:        {key2}")
    
    encrypted2 = encrypt(plaintext2, key2)
    decrypted2 = decrypt(encrypted2, key2)
    
    print(f"Encrypted:  {encrypted2}")
    print(f"Decrypted:  {decrypted2}")
    print(f"Success:    {decrypted2 == plaintext2}")

File: c/src/classical/transposition/simple/scytale.c
#include "cryptology/classical/transposition/simple/scytale.h"
#include "cryptology/classical/transposition/utf8_helpers.h"
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

int scytale_encrypt(const char *plaintext, int key, char *output, size_t output_size) {
    if (!plaintext || !output || key <= 0 || output_size == 0) {
        return -1;
    }
    
    // Clean and normalize text (UTF-8 aware, supports Turkish)
    char *clean_input = malloc(strlen(plaintext) * 3 + 1);  // Extra space for UTF-8
    if (!clean_input) return -1;
    
    if (clean_utf8_text(plaintext, clean_input, strlen(plaintext) * 3 + 1) != 0) {
        free(clean_input);
        return -1;
    }
    
    if (strlen(clean_input) == 0) {
        output[0] = '\0';
        free(clean_input);
        return 0;
    }
    
    size_t text_len = strlen(clean_input);
    size_t rows = (text_len + key - 1) / key;  // Ceiling division
    
    // Create matrix
    char **matrix = malloc(rows * sizeof(char*));
    if (!matrix) {
        free(clean_input);
        return -1;
    }
    
    for (size_t i = 0; i < rows; i++) {
        matrix[i] = malloc(key * sizeof(char));
        if (!matrix[i]) {
            // Clean up allocated memory
            for (size_t j = 0; j < i; j++) {
                free(matrix[j]);
            }
            free(matrix);
            free(clean_input);
            return -1;
        }
    }
    
    // Fill matrix row by row
    size_t text_index = 0;
    for (size_t row = 0; row < rows; row++) {
        for (size_t col = 0; col < key; col++) {
            if (text_index < text_len) {
                matrix[row][col] = clean_input[text_index++];
            } else {
                matrix[row][col] = ' ';  // Padding
            }
        }
    }
    
    // Read column by column
    size_t output_index = 0;
    for (size_t col = 0; col < key && output_index < output_size - 1; col++) {
        for (size_t row = 0; row < rows && output_index < output_size - 1; row++) {
            output[output_index++] = matrix[row][col];
        }
    }
    output[output_index] = '\0';
    
    // Clean up
    for (size_t i = 0; i < rows; i++) {
        free(matrix[i]);
    }
    free(matrix);
    free(clean_input);
    
    return 0;
}

int scytale_decrypt(const char *ciphertext, int key, char *output, size_t output_size) {
    if (!ciphertext || !output || key <= 0 || output_size == 0) {
        return -1;
    }
    
    // Clean the ciphertext (UTF-8 aware, supports Turkish)
    char *clean_input = malloc(strlen(ciphertext) * 3 + 1);  // Extra space for UTF-8
    if (!clean_input) return -1;
    
    if (clean_utf8_text(ciphertext, clean_input, strlen(ciphertext) * 3 + 1) != 0) {
        free(clean_input);
        return -1;
    }
    
    if (strlen(clean_input) == 0) {
        output[0] = '\0';
        free(clean_input);
        return 0;
    }
    
    size_t text_len = strlen(clean_input);
    size_t rows = text_len / key;
    if (text_len % key != 0) {
        rows++;
    }
    
    // Create matrix
    char **matrix = malloc(rows * sizeof(char*));
    if (!matrix) {
        free(clean_input);
        return -1;
    }
    
    for (size_t i = 0; i < rows; i++) {
        matrix[i] = malloc(key * sizeof(char));
        if (!matrix[i]) {
            // Clean up allocated memory
            for (size_t j = 0; j < i; j++) {
                free(matrix[j]);
            }
            free(matrix);
            free(clean_input);
            return -1;
        }
    }
    
    // Fill matrix column by column
    size_t text_index = 0;
    for (size_t col = 0; col < key; col++) {
        for (size_t row = 0; row < rows; row++) {
            if (text_index < text_len) {
                matrix[row][col] = clean_input[text_index++];
            }
        }
    }
    
    // Read row by row
    size_t output_index = 0;
    for (size_t row = 0; row < rows && output_index < output_size - 1; row++) {
        for (size_t col = 0; col < key && output_index < output_size - 1; col++) {
            output[output_index++] = matrix[row][col];
        }
    }
    output[output_index] = '\0';
    
    // Clean up
    for (size_t i = 0; i < rows; i++) {
        free(matrix[i]);
    }
    free(matrix);
    free(clean_input);
    
    return 0;
}

void scytale_get_key_range(size_t text_length, int *min_key, int *max_key) {
    if (text_length <= 0) {
        *min_key = 0;
        *max_key = 0;
    } else {
        *min_key = 1;
        *max_key = (int)text_length;
    }
}
